File: backend/scripts/check_env.py
#!/usr/bin/env python3
import sys
from pathlib import Path
from typing import Dict, List, Tuple

BACKEND_ENV_REQUIRED = {
    "ENV": {
        "description": "Environment mode (development/production)",
        "allowed_values": ["development", "production"],
        "default": "development",
    },
    "LOG_LEVEL": {
        "description": "Logging level (DEBUG/INFO/WARNING/ERROR)",
        "allowed_values": ["DEBUG", "INFO", "WARNING", "ERROR"],
        "default": "DEBUG",
    },
    "ALLOWED_ORIGINS": {
        "description": "Comma-separated CORS origins",
        "allowed_values": None,
        "default": "http://localhost:3000",
    },
}


def load_env_file(env_path: Path) -> Dict[str, str]:
    env_vars = {}
    if not env_path.exists():
        return env_vars
    with open(env_path) as f:
        for line in f:
            line = line.strip()
            if not line or line.startswith("#"):
                continue
            if "=" in line:
                key, value = line.split("=", 1)
                env_vars[key.strip()] = value.strip()
    return env_vars


def validate_env_vars(
    env_vars: Dict[str, str], required: Dict[str, Dict], env_file: str
) -> Tuple[bool, List[str]]:
    issues = []
    for key, config in required.items():
        if key not in env_vars:
            issues.append(
                f"Missing {key} in {env_file}\nDescription: {config['description']}\nDefault: {config['default']}"
            )
            continue
        value = env_vars[key]
        allowed = config.get("allowed_values")
        if not value:
            issues.append(f"Variable {key} in {env_file} is empty")

        if allowed and value not in allowed:
            issues.append(
                f"Invalid value for {key} in {env_file}\nCurrent: {value}\nAllowed: {', '.join(allowed)}"
            )
        return len(issues) == 0, issues


def check_backend(repo_root: Path) -> bool:
    env_file = repo_root / ".env"
    if not env_file.exists():
        print(f"No .env file found at {env_file}")
        return False
    env_vars = load_env_file(env_file)
    is_valid, issues = validate_env_vars(env_vars, BACKEND_ENV_REQUIRED, ".env")
    if is_valid:
        print("All required variables are set correctly")
        return True
    for issue in issues:
        print(issue)
    return False


def main():
    repo_root = Path(__file__).parent.parent
    backend_ok = check_backend(repo_root)
    sys.exit(0 if backend_ok else 1)


if __name__ == "__main__":
    main()

File: frontend/scripts/check-env.js
#!/usr/bin/env node
/* eslint-disable @typescript-eslint/no-require-imports */

/**
 * Environment Configuration Validator (Frontend)
 * Checks for missing or incorrectly configured environment variables.
 *
 * Usage:
 *     node scripts/check-env.js
 *     pnpm check-env
 */

const fs = require("fs");
const path = require("path");

// ============================================================================
// Configuration
// ============================================================================

const FRONTEND_ENV_REQUIRED = {
  NODE_ENV: {
    description: "Node environment (development/production)",
    allowedValues: ["development", "production"],
    default: "development",
  },
  NEXT_PUBLIC_API_URL: {
    description: "Backend API URL (must be accessible from browser)",
    allowedValues: null, // Any URL acceptable
    default: "http://127.0.0.1:8000",
  },
};

// ============================================================================
// Helper Functions
// ============================================================================

function loadEnvFile(envPath) {
  /**Load environment variables from .env.local file.*/
  const envVars = {};

  if (!fs.existsSync(envPath)) {
    return envVars;
  }

  const content = fs.readFileSync(envPath, "utf-8");
  for (const line of content.split("\n")) {
    const trimmed = line.trim();
    // Skip comments and empty lines
    if (!trimmed || trimmed.startsWith("#")) {
      continue;
    }
    // Parse KEY=VALUE
    if (trimmed.includes("=")) {
      const [key, ...valueParts] = trimmed.split("=");
      envVars[key.trim()] = valueParts.join("=").trim();
    }
  }

  return envVars;
}

function validateEnvVars(envVars, required, envFile) {
  /**Validate environment variables against requirements.*/
  const issues = [];

  for (const [key, config] of Object.entries(required)) {
    if (!(key in envVars)) {
      issues.push(
        `❌ Missing ${key} in ${envFile}\n` +
          `   Description: ${config.description}\n` +
          `   Default: ${config.default}`
      );
      continue;
    }

    const value = envVars[key];
    const allowed = config.allowedValues;

    // Validate against allowed values if specified
    if (allowed && !allowed.includes(value)) {
      issues.push(
        `⚠️  Invalid value for ${key} in ${envFile}\n` +
          `   Current: ${value}\n` +
          `   Allowed: ${allowed.join(", ")}`
      );
    }

    // Check for empty values
    if (!value) {
      issues.push(`❌ ${key} in ${envFile} is empty`);
    }
  }

  return [issues.length === 0, issues];
}

// ============================================================================
// Main Validation Logic
// ============================================================================

function checkFrontend(repoRoot) {
  /**Check frontend environment configuration.*/
  console.log("\n" + "=".repeat(70));
  console.log("🎨 FRONTEND ENVIRONMENT CHECK");
  console.log("=".repeat(70));

  const frontendPath = path.join(repoRoot, "frontend");
  const envFile = path.join(frontendPath, ".env.local");

  if (!fs.existsSync(envFile)) {
    console.log(`❌ No .env.local file found at ${envFile}`);
    console.log(`   Create one from .env.local.example:`);
    console.log(`   cp ${frontendPath}/.env.local.example ${frontendPath}/.env.local`);
    return false;
  }

  console.log(`✅ Found ${envFile}`);

  const envVars = loadEnvFile(envFile);
  const [isValid, issues] = validateEnvVars(envVars, FRONTEND_ENV_REQUIRED, ".env.local");

  if (isValid) {
    console.log("✅ All required variables are set correctly");
    console.log("\nConfiguration:");
    for (const [key, config] of Object.entries(FRONTEND_ENV_REQUIRED)) {
      const value = envVars[key] || config.default;
      console.log(`  - ${key}: ${value}`);
    }
    return true;
  } else {
    console.log("❌ Configuration issues found:\n");
    for (const issue of issues) {
      console.log(`  ${issue}\n`);
    }
    return false;
  }
}

function main() {
  /**Main entry point.*/
  const repoRoot = path.dirname(path.dirname(__dirname));

  // Parse CLI arguments
  if (process.argv.length > 2) {
    const arg = process.argv[2];
    if (arg === "--help" || arg === "-h") {
      console.log(
        "Environment Configuration Validator (Frontend)\n" +
          "Usage: node scripts/check-env.js\n" +
          "       pnpm check-env"
      );
      process.exit(0);
    } else {
      console.log(`Unknown argument: ${arg}`);
      process.exit(1);
    }
  }

  // Run checks
  const frontendOk = checkFrontend(repoRoot);

  // Summary
  console.log("\n" + "=".repeat(70));
  console.log("📋 SUMMARY");
  console.log("=".repeat(70));

  if (frontendOk) {
    console.log("✅ Frontend environment configuration is valid!");
    console.log("\n🚀 You can now start development:");
    console.log("   pnpm dev");
    process.exit(0);
  } else {
    console.log("❌ Fix the issues above before continuing");
    process.exit(1);
  }
}

main();
